### ISSUE TITLE
determine if a run is complete before starting the demultiplexing

### ISSUE THREAD

**Issue by tcezard** [EdinburghGenomics/Analysis-Driver#215]
When the data copy fails between the sequencer and our storage, some bcl files ends up being corrupted. bcl2fastq fails and we need to identify all the offending bcls by running bcl2fastq wit ```--ignore-missing-bcl```.
This is a very slow process and in order to gain some time we should check that all the bcl files are there and valid.
In order to do that we can only read the bcl files header [see example](https://github.com/EdinburghGenomics/well_duplicates/blob/master/bcl_direct_reader.py#L187) which should be relatively fast
We then can report all the corrupted files and raise an error

**Comment by mwhamgenomics** [comment]
It's possible to replicate the above example from WellDuplicates, however it takes ~20 mins per lane, or about 2 hrs 40 mins in total. Since it's quite a big job, we should probably run it through `executor`.

We could split the BCL check into its own project/dependency. Alternatively if we don't want to do that, we could just have it as a command run through `executor`, but then it wouldn't be tested/versioned.

**Comment by mwhamgenomics** [comment]
CProfile report from pure-Python bcl validation:
```
   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.549    0.549 1313.856 1313.856 bcl_validation.py:15(validate_bcl_files)
    29760    0.242    0.000  934.839    0.031 lib/python3.4/gzip.py:349(read)
    29760    0.269    0.000  934.502    0.031 lib/python3.4/gzip.py:425(_read)
    89376  933.363    0.010  933.363    0.010 {method 'read' of '_io.BufferedReader' objects}
    29760    0.296    0.000  930.662    0.031 lib/python3.4/gzip.py:291(_read_gzip_header)
    89280    0.222    0.000  930.270    0.010 lib/python3.4/gzip.py:79(read)
       98    0.002    0.000  363.044    3.705 lib/python3.4/site-packages/egcg_core/util.py:11(find_files)
       98    0.017    0.000  362.990    3.704 lib/python3.4/glob.py:9(glob)
59818/29962    0.164    0.000  362.972    0.012 lib/python3.4/glob.py:20(iglob)
    29760    0.091    0.000  362.229    0.012 lib/python3.4/glob.py:75(glob0)
    29760    0.064    0.000  362.022    0.012 lib/python3.4/posixpath.py:168(lexists)
    29760  361.958    0.012  361.958    0.012 {built-in method lstat}
```

Traversing the folders is slow as expected, but most of the slowdown seems to be in `gzip.read` and `gzip._read_gzip_header`.

**Comment by tcezard** [comment]
Another option to do this validation before we run the pipeline, would be to use zlib directly or to write C code. Both options are quite unappealing ...

We could also reinstate the Run Processing from the start of the run with an infinite while loop. instead of copying the run to a remote location we would check every bcl file as it appear on the disk.

**Comment by mwhamgenomics** [comment]
We tried decompressing directly through `zlib`, but it still took ~20 mins per lane, so it looks like the rate-limiting step is the filesystem.

If we want to run bcl validation, we can stagger it with the sequencing run:
- kick off the pipeline as soon as the run folder appears
- repeatedly scan for bcls and validate
- record checked bcls to a file to avoid checking the same file twice
- scan until RTAComplete.txt exists, or Lims returns 'RunCompleted', 'RunFailed', etc.

- we would also be able to parse the sample sheet to the reporting app immediately

**Comment by mwhamgenomics** [comment]
After some discussion, if the bcl header validation fails, then that means something has gone very wrong and the sequencer run is probably bad. As far as I know, corrupted bcls have always been due to problems in transfer, resulting in the file being truncated/missing.

We could just check for general file integrity, i.e. it exists, is non-zero length and is valid gzip.

**Comment by mwhamgenomics** [comment]
This was closed in #218.

**Comment by mwhamgenomics** [comment]
Actually, let's keep this open until the pipeline starts calling it.

**Comment by mwhamgenomics** [comment]
We should make sure multiple compute nodes can safely write to the same file in parallel.

**Comment by mwhamgenomics** [comment]
This was closed in #230.
